hoomd/device.py

```diff
@@ -1,51 +1,37 @@
-# Copyright (c) 2009-2019 The Regents of the University of Michigan
-# This file is part of the HOOMD-blue project, released under the BSD 3-Clause License.
+# Copyright (c) 2009-2019 The Regents of the University of Michigan This file is
+# part of the HOOMD-blue project, released under the BSD 3-Clause License.
 
-# Maintainer: tommy-waltmann / All Developers are free to add commands for new features
-
-r""" Devices available to run simulations
-
-A device object represents the hardware (whether CPU, GPU, or Auto) the simulation will run on. Creating a device
-object will automatically add it to the simulation context. A device in mode Auto is chosen by default for the user,
-but they can chose a new one before starting the simulation. Devices in mode Auto will choose GPU at runtime if available,
-otherwise simulations will run on the CPU.
-
-Note:
-    Device objects have the following properties:
-
-
-Attributes:
-    mode (str): gpu or cpu
-    comm (:py:mod:`hoomd.comm.Communicator`): communicator object held by this device
-    num_threads (int): the number of CPU threads to be used in simulation, settable
-    gpu_ids (list(int)): list of names of the gpus, if in gpu mode
-    num_ranks (int): the number of ranks
-    notice_level (int): minimum level of notice messages to print, settable
-    memory_tracback (bool): If true, enable memory allocation tracking (*only for debugging/profiling purposes*)
-    gpu_error_checking (bool): Whether or not CUDA error checking is enabled, settable.
-    msg_file (str): The name of the file to write messages. Set this property to None to write to stdout/stderr.
-"""
+"""Choose which hardware device(s) should execute the simulation."""
 
 import contextlib
 import os
-import time
+import warnings
 import hoomd
 from hoomd import _hoomd
 
-class _device(hoomd.meta._metadata):
 
-    def __init__(self, communicator, notice_level, msg_file, shared_msg_file):
+class _Device:
+    """Base class device object.
+
+    Provides methods and properties common to `CPU` and `GPU`.
+
+    .. rubric:: TBB threads
 
-        # metadata stuff
-        hoomd.meta._metadata.__init__(self)
-        self.metadata_fields = ['gpu_ids', 'mode', 'num_ranks']
-        if _hoomd.is_TBB_available():
-            self.metadata_fields.append('num_threads')
+    Set `num_cpu_threads` to `None` and TBB will auto-select the number of CPU
+    threads to execute. If the environment variable ``OMP_NUM_THREADS`` is set,
+    HOOMD will use this value. You can also set `num_cpu_threads` explicitly.
+
+    Note:
+        At this time **very few** features in HOOMD use TBB for threading.
+        Most users should employ MPI for parallel simulations.
```

Comment:
We should add a warning for users not to instantiate this class.

Reply:
done.